File: python/circular-buffer/circular_buffer.py
class BufferFullException(BufferError):
    pass

class BufferEmptyException(BufferError):
    pass


class CircularBuffer(object):
    def __init__(self, capacity):
        self.header = Header(capacity)
        self.buffer = [None] * capacity

    def is_empty(self):
        return self.buffer[self.header.read] is None

    def is_full(self):
        return self.header.read == self.header.write \
            and not self.is_empty()

    def read(self):
        if self.is_empty():
            raise BufferEmptyException('Circular buffer is empty')
        data = self.buffer[self.header.read]
        self.buffer[self.header.read] = None
        self.header.inc_r()
        return data

    def write(self, data):
        if self.is_full():
            raise BufferFullException('Circular buffer is full')
        self.buffer[self.header.write] = data
        self.header.inc_w()
        return self

    def overwrite(self, data):
        if self.is_full():
            self.read()
        return self.write(data)

    def clear(self):
        self.buffer = [None] * len(self.buffer)
        self.header.reset()
        return self


class Header(object):
    def __init__(self, capacity):
        self.cap = capacity
        self.reset()

    def reset(self):
        self.p = {'r': 0, 'w': 0}

    @property
    def read(self):
        return self.p['r']

    @property
    def write(self):
        return self.p['w']

    def inc_r(self, header='r'):
        self.p[header] = (self.p[header] + 1) % self.cap

    def inc_w(self):
        self.inc_r('w')

File: elixir/circular-buffer/lib/circular_buffer.ex
defmodule CircularBuffer do
  @moduledoc """
  An API to a stateful process that fills and empties a circular buffer
  """

  use Agent

  defstruct [:capacity, data: [], count: 0]

  @doc """
  Create a new buffer of a given capacity
  """
  @spec new(capacity :: integer) :: {:ok, pid}
  def new(cap) do
    Agent.start_link(fn -> %CircularBuffer{capacity: cap} end)
  end

  @doc """
  Read the oldest entry in the buffer, fail if it is empty
  """
  @spec read(buffer :: pid) :: {:ok, any} | {:error, atom}
  def read(buffer) do
    Agent.get_and_update(buffer, fn state ->
      if state.count == 0 do
        {{:error, :empty}, state}
      else
        {{:ok, hd(state.data)},
         state
         |> Map.update!(:data, &tl/1)
         |> Map.update!(:count, &(&1 - 1))}
      end
    end)
  end

  @doc """
  Write a new item in the buffer, fail if is full
  """
  @spec write(buffer :: pid, item :: any) :: :ok | {:error, atom}
  def write(buffer, item) do
    Agent.get_and_update(buffer, fn state ->
      if state.count == state.capacity do
        {{:error, :full}, state}
      else
        {:ok,
         state
         |> Map.update!(:data, &(&1 ++ [item]))
         |> Map.update!(:count, &(&1 + 1))}
      end
    end)
  end

  @doc """
  Write an item in the buffer, overwrite the oldest entry if it is full
  """
  @spec overwrite(buffer :: pid, item :: any) :: :ok
  def overwrite(buffer, item) do
    case write(buffer, item) do
      {:error, :full} ->
        read(buffer)
        write(buffer, item)

      result ->
        result
    end
  end

  @doc """
  Clear the buffer
  """
  @spec clear(buffer :: pid) :: :ok
  def clear(buffer) do
    Agent.update(buffer, fn state -> %CircularBuffer{capacity: state.capacity} end)
  end
end
